REGRESSION(291600@main): TestWebKitAPI.GPUProcess.RelaunchOnCrash is a constant failure

achristensen07 opened the pull request:
#### 4b60d193835eaf5c2071fa9c6c104bb24503d162
<pre>
REGRESSION(291600@main): TestWebKitAPI.GPUProcess.RelaunchOnCrash is a constant failure
<a href="https://bugs.webkit.org/show_bug.cgi?id=289469">https://bugs.webkit.org/show_bug.cgi?id=289469</a>
<a href="https://rdar.apple.com/146666667">rdar://146666667</a>

Reviewed by Youenn Fablet.

In this test, RemoteAudioDestinationProxy::startRenderingThread is called twice.
The second time, offThreadRendering hits the &quot;break&quot; statement because m_shouldStopThread
was not being reset after RemoteAudioDestinationProxy::gpuProcessConnectionDidClose called
stopRenderingThread.

When audio playing stops, AudioContext::isPlayingAudioDidChange dispatches to the main thread
to tell it to tell the UI process that the media state changed through WebPage::isPlayingMediaDidChange.
291340@main made it a race condition between the UI process being informed of the new GPU process
and the UI process being informed of the audio stopping.  291600@main subtly changed the timing
of main thread behavior, especially in debug builds, which made the audio lose the race more often.

* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.cpp:
(WebKit::RemoteAudioDestinationProxy::gpuProcessConnectionDidClose):

Canonical link: <a href="https://commits.webkit.org/292265@main">https://commits.webkit.org/292265@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b2bccedd54d487478c50c5603f585e8fe0026f7f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95441 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15044 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/4915 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/100491 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/45947 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15331 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23477 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/100491 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30063 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98444 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/11476 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86163 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/100491 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11187 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45283 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/81369 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/3987 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/102528 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22491 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/16431 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/81814 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/22740 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82171 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/81186 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25762 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3190 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/15814 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15336 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22460 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/27599 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22119 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/25594 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/23859 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf reviewed (approved):
LGTM.
Maybe we should set `m_shouldStopThread ` back to false in `RemoteAudioDestinationProxy::stopRenderingThread` just after waiting for completion of the waiting thread.

achristensen07 committed:
REGRESSION(291600@main): TestWebKitAPI.GPUProcess.RelaunchOnCrash is a constant failure
https://bugs.webkit.org/show_bug.cgi?id=289469
rdar://146666667

Reviewed by Youenn Fablet.

In this test, RemoteAudioDestinationProxy::startRenderingThread is called twice.
The second time, offThreadRendering hits the "break" statement because m_shouldStopThread
was not being reset after RemoteAudioDestinationProxy::gpuProcessConnectionDidClose called
stopRenderingThread.

When audio playing stops, AudioContext::isPlayingAudioDidChange dispatches to the main thread
to tell it to tell the UI process that the media state changed through WebPage::isPlayingMediaDidChange.
291340@main made it a race condition between the UI process being informed of the new GPU process
and the UI process being informed of the audio stopping.  291600@main subtly changed the timing
of main thread behavior, especially in debug builds, which made the audio lose the race more often.

* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.cpp:
(WebKit::RemoteAudioDestinationProxy::gpuProcessConnectionDidClose):

Canonical link: https://commits.webkit.org/292265@main
